src/mo_frontend/typing.ml

```diff
@@ -1872,9 +1872,16 @@ and infer_pat' env pat : T.typ * Scope.val_env =
         "pattern branches have incompatible types,\nleft consumes%a\nright consumes%a"
         display_typ_expand t1
         display_typ_expand t2;
-    if ve1 <> T.Env.empty || ve2 <> T.Env.empty then
-      error env pat.at "M0105" "variables are not allowed in pattern alternatives";
-    t, T.Env.empty
+    if T.Env.keys ve1 <> T.Env.keys ve2 then
+      error env pat.at "M0184" "different set of bindings in pattern alternatives";
+    let check_same_bind_type bind t1 t2 = if not (T.eq t1 t2) then
+      local_error env pat.at "M0185" "types for alternative pattern variables %s differ%a%a"
```

Comment:
this should be `error`, since it makes checking the function's body impossible. Also need to spit the `fail/` test up in parts.

Reply:
see 130c51429 (and parent)